### PR TITLE
Fix time overlap handling in concatenation

### LINKED ISSUE #2248: Concatenate implementation is broken
The function `_check_time_overlaps` [first constructs a list of time coordinates from the given cubes](https://github.com/ESMValGroup/ESMValCore/blob/f39d14a502ba92da0219ab8433453e1c90d1ea46/esmvalcore/preprocessor/_io.py#L211), then does an [index loop over those](https://github.com/ESMValGroup/ESMValCore/blob/f39d14a502ba92da0219ab8433453e1c90d1ea46/esmvalcore/preprocessor/_io.py#L212), and in the loop body removes cubes based on the index from the cube list.

This way, times are checked that do not correspond to the cube under question and the method can produce wrong results for three or more cubes.

### PR DESCRIPTION
**Pull request by zklaus**
## Description

Closes #2248 

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: esmvalcore/preprocessor/_io.py [hunk 1]
```diff
@@ -206,70 +205,96 @@ def _concatenate_cubes(cubes, check_level):
...
+    current_cube = _TrackedCube.from_cube(cubes[0])
+    for new_cube in map(_TrackedCube.from_cube, cubes[1:]):
+        if new_cube.start > current_cube.end:
+            # no overlap, use current cube and start again from new cube
+            logger.debug("Using %s", current_cube.cube)
+            new_cubes.append(current_cube.cube)
+            current_cube = new_cube
+            continue
+        # overlap
+        if current_cube.end > new_cube.end:
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_io.py:265]:
```suggestion
        if current_cube.end >= new_cube.end:
```
?

Comment (zklaus) [reply esmvalcore/preprocessor/_io.py:265]: Yeah, I had it like that first, but then https://github.com/ESMValGroup/ESMValCore/blob/f39d14a502ba92da0219ab8433453e1c90d1ea46/tests/integration/preprocessor/_io/test_concatenate.py#L393 fails. Honestly, the semantics of this function are a bit dubious, but this PR only addresses the implementation.

**Comment by zklaus** [comment]
The coverage fails only due to one line, the special case when the passed cube list is shorter than two (i.e. has zero or one entries), in which case the function returns the input cube list unmodified. This is not covered by tests because the test works at the level of the `concatenate` function, which performs this check before, thus even though `concatenate` *is* tested with an empty cube list, this does not reach `_check_time_overlaps`.

There are two possible solutions: Add a test for `_check_time_overlaps` itself or remove the check inside the function, taking it as a guarantee by the caller. A third approach is to ignore the code coverage deterioration.

What do the reviewers think?

**Review by bouweandela — APPROVED** [review]
Looks good to me!

**Comment by valeriupredoi** [comment]
well you can mock the concatenate func with that single cube, but nay bother, if Bouwe is OK with that am too :+1:

**Comment by valeriupredoi** [comment]
plus 97.87% is pretty much 100% in my books :grin:

**Pull request merged**